src/main/java/htsjdk/variant/vcf/AbstractVCFCodec.java

```diff
@@ -750,6 +750,56 @@ private static boolean isVCFStream(final InputStream stream, final String MAGIC_
         return new LazyGenotypesContext.LazyData(genotypes, header.getSampleNamesInOrder(), header.getSampleNameToOffset());
     }
 
+    private Object genotypeValsToTargetType(final String key, final String vals) {
+        if (!header.hasFormatLine(key)) {
+            return vals;
+        }
+
+        final VCFFormatHeaderLine formLine = header.getFormatHeaderLine(key);
+        final VCFHeaderLineType valType = formLine.getType();
+        final boolean isSingleVal = formLine.isFixedCount()
+                ? formLine.getCount() == 1
+                : !vals.contains(VCFConstants.GENOTYPE_ATTRIBUTE_VALUE_SEPARATOR);
+
+        if (isSingleVal) {
+            return valType == VCFHeaderLineType.String
+                    ? vals
+                    : getMapperFunction(valType).apply(vals);
+        } else {
+            final List<?> targetTypeVals = mapToFormatLineType(
+                    ParsingUtils.split(vals, VCFConstants.GENOTYPE_ATTRIBUTE_VALUE_SEPARATOR_CHAR),
+                    valType);
+            return targetTypeVals.isEmpty() ? vals : targetTypeVals;
+        }
+    }
+
+    private List<?> mapToFormatLineType(final List<String> genVals, final VCFHeaderLineType genValType) {
+        if (genValType == VCFHeaderLineType.String) {
+            return genVals;
+        }
+
+        final List<Object> genRealTypeVals = new ArrayList<>(genVals.size());
+        final Function<String, Object> mapperFunc = getMapperFunction(genValType);
+
+        genVals.forEach(genVal -> {
+            if (!genVal.equals(VCFConstants.MISSING_VALUE_v4)) {
+                genRealTypeVals.add(mapperFunc.apply(genVal));
+            }
```

Comment:
something else should happen if there's a missing value in the middle of a list of strings....so "0,1,.,2" should NOT convert to [0,1,2] but rather something like [0, 1, null, 2] or [0, 1, MISSING_INT_VALUE, 2]